From 6db3682d70e3e9dce7b9e32ddfd94378773f1d38 Mon Sep 17 00:00:00 2001
From: Felipe Contreras <felipe.contreras@gmail.com>
Date: Thu, 20 May 2021 16:26:25 -0500
Subject: [PATCH] Add Gem.default_install

The configuration Gem.default_user_install should have never been added,
as the same can be achieved by specifying the default install directory,
this is what the new Gem.default_install does.

This flag should be used to set Gem.home so it helps everyone, not just
the install command.

By calling Gem.default_user_install in the right place, bundler can also
make use of this, and other tools like `gem env` show the correct
location where gems are going to be installed to.

Gem.default_dir is where the system gems are installed to,
Gem.default_install is where the gems installed by the user are
installed to.

This allows for more flexibility, for example allow to specify another
directory, or another directory only if the user is root.

The current behavior remains unaffected: gems are still installed to
based on Gem.default_user_install by default.

Signed-off-by: Felipe Contreras <felipe.contreras@gmail.com>
---
 lib/rubygems/defaults.rb            | 14 ++++++++------
 lib/rubygems/path_support.rb        |  2 +-
 test/rubygems/test_gem_installer.rb |  2 --
 3 files changed, 9 insertions(+), 9 deletions(-)

diff --git a/lib/rubygems/defaults.rb b/lib/rubygems/defaults.rb
index db07681a17..ccc483b11e 100644
--- a/lib/rubygems/defaults.rb
+++ b/lib/rubygems/defaults.rb
@@ -9,6 +9,13 @@ module Gem
   @pre_uninstall_hooks  ||= []
   @pre_install_hooks    ||= []
 
+  ##
+  # Determines the default install directory
+
+  def self.default_install
+    Gem.default_user_install ? Gem.user_dir : Gem.default_dir
+  end
+
   ##
   # An Array of the default sources that come with RubyGems
 
@@ -239,12 +246,7 @@ def self.default_cert_path
   # Enables automatic installation into user directory
 
   def self.default_user_install # :nodoc:
-    if !ENV.key?("GEM_HOME") && File.exist?(Gem.dir) && !File.writable?(Gem.dir)
-      Gem.ui.say "Defaulting to user installation because default installation directory (#{Gem.dir}) is not writable."
-      return true
-    end
-
-    false
+    File.exist?(Gem.default_dir) && !File.writable?(Gem.default_dir)
   end
 
   ##
diff --git a/lib/rubygems/path_support.rb b/lib/rubygems/path_support.rb
index 13091e29ba..455f413f3d 100644
--- a/lib/rubygems/path_support.rb
+++ b/lib/rubygems/path_support.rb
@@ -24,7 +24,7 @@ class Gem::PathSupport
   # hashtable, or defaults to ENV, the system environment.
   #
   def initialize(env)
-    @home = normalize_home_dir(env["GEM_HOME"] || Gem.default_dir)
+    @home = normalize_home_dir(env["GEM_HOME"] || Gem.default_install)
     @path = split_gem_path env["GEM_PATH"], @home
 
     @spec_cache_dir = env["GEM_SPEC_CACHE"] || Gem.default_spec_cache_dir
diff --git a/test/rubygems/test_gem_installer.rb b/test/rubygems/test_gem_installer.rb
index 34415aa7dd..c704441dcf 100644
--- a/test/rubygems/test_gem_installer.rb
+++ b/test/rubygems/test_gem_installer.rb
@@ -1989,7 +1989,6 @@ def test_process_options_fallback_to_user_install_when_gem_home_not_writable
     installer = use_ui(@ui) { Gem::Installer.at @gem }
 
     assert_equal Gem.user_dir, installer.gem_home
-    assert_equal "Defaulting to user installation because default installation directory (#{@gemhome}) is not writable.", @ui.output.strip
   ensure
     FileUtils.chmod 0o755, @gemhome
     ENV["GEM_HOME"] = orig_gem_home
@@ -2010,7 +2009,6 @@ def test_process_options_does_not_fallback_to_user_install_when_gem_home_not_wri
     installer = use_ui(@ui) { Gem::Installer.at @gem, user_install: false }
 
     assert_equal @gemhome, installer.gem_home
-    assert_empty @ui.output.strip
   ensure
     FileUtils.chmod 0o755, @gemhome
     ENV["GEM_HOME"] = orig_gem_home
-- 
0.1

